design/proposals/variable_schema_validation.md

```diff
@@ -0,0 +1,264 @@
+      - [Input Validation](#input-validation)
+      - [Secrets Encryption](#secrets-encryption)
+      - [Builtin Schemas](#builtin-schemas)
+      - [Additional Required Functionality](#additional-required-functionality)
+  - [Main Advantages of This Proposal](#main-advantages-of-this-proposal)
+  - [Affected Areas](#affected-areas)
+  - [Backwards Compatibility](#backwards-compatibility)
+  - [Performance](#performance)
+  - [Security](#security)
+  - [Documentation](#documentation)
+  - [Version Update](#version-update)
+  - [Future Work](#future-work)
+  - [Delivery Plan](#delivery-plan)
+  
+## Introduction
+
+Conjur variables are resources that allow a user or an application to define and store sensitive information.
+
+In many cases, such sensitive information is comprised of multiple pieces that are tightly coupled and together represent a complete object. A few examples:
+
+- In order to access a MySQL database, a user needs a username, a password and an address, URL or connection string.
+- In order to access a REST API exposed by a certain service, a user might need either a username and password, a private key and a public certificate, or a token. In addition, the user will need the server public certificate and its address.
+
+Currently, this is handled by defining multiple variables in Conjur. Each variable contains a single value, such as a username, a password, etc. To bundle these pieces together, we use a naming convention. For example:
+
+```yaml
+- !policy
+  id: my-policy
+  body:
+    - &vars
+      - !variable mysql-db-creds/password
+      - !variable mysql-db-creds/username
+      - !variable mysql-db-creds/address
+      - !variable mysql-db-creds/port
+
+    - !host my-app
+    
+    - !permit
+      role: !host my-app
+      privileges: [ read,execute ]
+      resource: *vars
+```
+
+## Proposed Solution
+
+The proposed solution is to add the ability to use JSON schemas and link them to variables, for which we want validation of their content. This will allow a single variable to store multiple, related values. While it's already possible today, to store JSON structured information in a variable, this proposed feature adds two added values:
+
+- The JSON structure is validated and enforced. If the input is invalid, the user will get a proper error message and the variable update will be denied.
+- The JSON structure will not be encrypted completely, but only the sensitive information. This will allow future seach capabilities based on the non sensitive information.
+
+### User Experience
+
+Let's have a look at the following example:
+
+```yaml
+- !policy
+  id: my-policy
+  body:
+  - !variable
+    id: mysql-db-creds
+    schema: !variable /conjur/schemas/mysql-schema
+
+  - !host my-app
+  
+  - !permit
+    role: !host my-app
+    privileges: [ read,execute ]
+    resource: !variable mysql-db-creds
+```
+
+And
+
+```yaml
+- !policy
+  id: conjur/schemas
+  body:
+  - &schemas
```

Comment:
this doesn't require any implementation, correct? this alias can be any word

Reply:
Correct